Fix SCP quorum reaching consensus

AndrejMitrovic opened the pull request:
There might still be some slight timing issues. If Travis/Github CI fails, I'll see about looking into this more.

One of the test-cases would fail only when the full test-suite ran, so it seems there have to be some delays between nomination rounds (after externalization). This is something I will look into more soon.

Partial fixes for https://github.com/bpfkorea/agora/issues/518

AndrejMitrovic commented:
Yeah it's already failing. But that's ok, it's likely a timing thing. I'll look into it.

TrustHenry commented on source/agora/node/Ledger.d:
```diff
@@ -299,10 +310,12 @@ public class Ledger
         if (txs.length != Block.TxsInBlock)
             return;  // not enough valid txs
 
+        this.is_nominating = true;
```
`scope(exit) this.is_nominating = false;`

Maybe should use it.

AndrejMitrovic commented:
Ok let me try to split this up, maybe one of the commits triggers an invalid state. The errors seem like segfaults / assertion failures in unexpected places.

```
Running ./build/agora-unittests 
terminate called after throwing an instance of 'std::invalid_argument'
  what():  invalid public key type
```

```
Running ./build/agora-unittests 
libc++abi.dylib: terminating with uncaught exception of type std::runtime_error: moved to a bad state (ballot protocol)
Program exited with code -6
```

AndrejMitrovic committed:
Fix the NetworkManager tests

The test needs to consist of one validator,
one full node (victim), and another full node (byzantine).

The Validator is the only node allowed to create blocks,
while the victim node tries to retrieve blocks via getBlocksFrom().

The victim will reject blocks from the byzantine node because
the blocks are not cryptographically valid.

AndrejMitrovic committed:
Don't alter Ledger state during nomination

If the Ledger's state is changed while a node is undergoing
a nomination process before externalization, this can and
will cause validateValue() in the SCP protocol to fail,
and will later cause the externalize message to be rejected
due to an unexpected Ledger state (wrong block height).

AndrejMitrovic committed:
Fix the Ledger catch-up phase test

The nodes were previously creating blocks on their own,
instead of retrieving them via getBlocksFrom (which was the test).

Changed the test to spawn only 1 validator node and 3 full-nodes,
ensuring those nodes only retrieve blocks via getBlocksFrom(),
and never create new ones themselves.

AndrejMitrovic committed:
Remove unused tryNominate duplicate method

tryNominateTxSet() can be used alone.

AndrejMitrovic committed:
Add OneValidator to NetworkTopology

This is useful to simulate cases where there is a single
validating node which creates blocks,
while the rest of the nodes are full nodes.

AndrejMitrovic committed:
Ensure Validators do not recursively nominate

In the current design, we do not use a timer to trigger nominations.
Instead, a new transaction being received can trigger the nomination
protocol if there are enough transactions in the pool to form a
new block.

There is a timing issue where a new transaction could be received
while there was already a nomination round in progress.

Fix the issue by temporarily adding a status flag,
which tells us whether we are currently in a nomination round.

This will later be replaced with the use of timers (real / virtual),
ensuring that a nomination round is started in predictable time segments.

AndrejMitrovic committed:
Do not nominate blocks if we're a full node

Only Validator nodes are allowed to create new blocks.

Also fixed the BanManager tests:

Previously the test was using putTransaction expecting the
banned node to create the blockchain on its own.

However this is a non-validating node, therefore it should only
use getBlocksFrom() to manipulate its Ledger state.

AndrejMitrovic commented:
@Geod24 I've had to temporarily remove the commit which sets the proper threshold to > 1 in the config, there is a failure on Linux platforms (but on the Mac it works fine).

With this, the PR is green and can be merged. In the meantime I'll continue debugging and get the Linux failure fixed.

Geod24 commented on source/agora/test/Ledger.d:
```diff
@@ -96,14 +96,15 @@ unittest
 
-    auto nodes = network.apis.values;
+    // node_1 is the validator
+    auto nodes = network.keys.map!(key => network.apis[key]).array;
```
Slight optimization: Since you call `array` at the end, `byKeys` would be better to avoid double allocation (but it's unittest code so heh)

AndrejMitrovic replied:
`.keys` here refers to the `keys` field inside of `TestAPIManager`, not the `.keys()` of a hashmap.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -261,6 +261,10 @@ public class Ledger
 
     public void tryNominateTXSet () @safe
```
Third commit ("Do not nominate blocks if we're a full node"): typo "agor atest-suite"

Geod24 commented on source/agora/test/Ledger.d:
```diff
@@ -63,6 +64,7 @@ unittest
 
         block_txes ~= txs.sort.array;
         last_txs = txs;
+        Thread.sleep(500.msecs);
```
Please add a comment as to why it's required

AndrejMitrovic replied:
Might not be necessary anymore, I'll try to remove it.

Geod24 commented on source/agora/test/BanManager.d:
```diff
@@ -79,42 +80,42 @@ unittest
+            retryFor(node.getBlockHeight() == 1 + (block_idx + 1),
+                4.seconds,
+                format("Block %s Node %s has block height %s. Expected: %s",
+                    block_idx + 1, idx, node.getBlockHeight().to!string, 1 + (block_idx + 1))));
```
Why the `to!string` here ?

AndrejMitrovic replied:
It used to be in place of the `format` call, I'll fix it.

Geod24 commented on source/agora/test/BanManager.d:
```diff
@@ -79,42 +80,42 @@ unittest
+    }
 
     // wait for node 3 to be banned and all putTransaction requests to time-out
     Thread.sleep(2.seconds);
```
Is this still relevant ?

AndrejMitrovic replied:
Yeah because it still takes actual real world time for a node to be banned.

I can see two ways of making this better:
- Use a virtual clock so we can "fast-forward" events. It's still tricky to write code for this. I've attempted this once.
- Add an event handling mechanism, so we can wait until some event fires. Something like this:

```D
class Events
{
    void onBannedNode (Node node) { ... }
    void onWhatever (...) { ... }
}
```

Then in the test-suite we could provide a subclass of `Events`, give that to the `Node`, and we could listen and wait for events. Instead of this `Thread.sleep(2.seconds);` call, we could just wait for the `onBannedNode` call.

Geod24 replied:
I wonder if anyone has implemented https://en.wikipedia.org/wiki/Server-sent_events in D

Geod24 commented on source/agora/test/BanManager.d:
```diff
@@ -79,42 +80,42 @@ unittest
+        retryFor(node.getBlockHeight() == 6,
+            4.seconds,
+            format("Node %s has block height %s. Expected: %s",
+                idx, node.getBlockHeight().to!string, 6)));
```
Ditto `to!striing`

Geod24 commented on source/agora/test/BanManager.d:
```diff
@@ -79,42 +80,42 @@ unittest
-    retryFor(node_3.getBlockHeight() == 1, 1.seconds, node_3.getBlockHeight().to!string);
-
-    // clear putTransaction filter
+    // sanity check: block height should not updated, node 3 is not a validator and cannot make blocks
```
Maybe add your explanation that the Ledger can only be manipulated through `getBlockFrom` ?

Geod24 commented on source/agora/test/NetworkManager.d:
```diff
@@ -122,37 +126,53 @@ unittest
+            retryFor(node.getBlockHeight() == block_idx + 1,
+                4.seconds,
+                format("Node %s has block height %s. Expected: %s",
+                    idx, node.getBlockHeight().to!string, block_idx + 1)));
```
`to!string`

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -297,11 +311,12 @@ public class Ledger
             return;  // not enough valid txs
 
         this.is_nominating = true;
+        scope (exit) is_nominating = false;
```
This should be in an earlier commit

AndrejMitrovic committed:
Remove unnecessary to!string calls.

AndrejMitrovic commented:
Fixed up.

Geod24 reviewed (approved):
Very nice!
